### ISSUE TITLE
Transforms and direct proxying

### ISSUE THREAD

**Issue by leastprivilege** [microsoft/reverse-proxy#801]
### What should we add or change to make your life better?
I am using this approach for a specialized scenario:

https://microsoft.github.io/reverse-proxy/articles/direct-proxying.html

Still I'd like to re-use your transforms "library". But I don't see an easy connection between the two. Am I missing something? Would a "standalone" transforms-builder be useful where I can pass the result in `ProxyAsync`?

### Why is this important to you?
I don't want to re-invent the wheel ;)

**Comment by Tratcher** [comment]
I have been wondering about this disparity. Which transforms did you have in mind? In a related discussion (https://github.com/microsoft/reverse-proxy/discussions/768) we found that many of the transforms provided minimal functionality and only existed to map config to code.

It is possible to re-use the structured transform infrastructure today, I think the interesting question is how much of it do you need and how hard is it to re-use those parts?

The transform types themselves are all public and can be new'd up and executed directly. That might make sense if you're only looking for one or two. If you want to run many then it might help if we made the [StructuredTransformer](https://github.com/microsoft/reverse-proxy/blob/45d063e06d88f4aace4e34f7ec8e1b50f16f53dc/src/ReverseProxy/Service/RuntimeModel/Transforms/StructuredTransformer.cs#L18-L26) class public?

The ITransformBuilder exists to map config based rules to their associated transform classes. That doesn't seem applicable for the IHttpProxy scenario?

**Comment by leastprivilege** [comment]
e.g. the forwarded headers...

Yes - the chaining of transforms should be public and independent of the config plumbing. 

Maybe same goes for the HTTP message invoker "factory" (IOW - returning a message invoke based on a name).

**Comment by Tratcher** [comment]
> Maybe same goes for the HTTP message invoker "factory" (IOW - returning a message invoke based on a name).

Yeah, some folks have asked about deriving from the internal ProxyHttpClientFactory. Would you want to use that together with IHttpProxy? The CreateClient [inputs](https://github.com/microsoft/reverse-proxy/blob/45d063e06d88f4aace4e34f7ec8e1b50f16f53dc/src/ReverseProxy/Service/Proxy/Infrastructure/ProxyHttpClientContext.cs#L14) are a little weird, but not prohibitive. I'll fork that out to its own issue.

**Comment by Tratcher** [comment]
Here's what it would look like if we made just the StructuredTransformer class public.

            var requestTransforms = new List<RequestTransform>();
            requestTransforms.Add(new RequestHeaderXForwardedForTransform("x-forwarded-for", append: true));
            requestTransforms.Add(new RequestHeaderXForwardedProtoTransform("x-forwarded-proto", append: true));
            requestTransforms.Add(new RequestHeaderXForwardedHostTransform("x-forwarded-host", append: true));

            var transformer =
                new StructuredTransformer(
                    copyRequestHeaders: default,
                    copyResponseHeaders: default,
                    copyResponseTrailers: default,
                    requestTransforms,
                    Array.Empty<ResponseTransform>(),
                    Array.Empty<ResponseTrailersTransform>());

That's.. meh.

The current experience for adding transforms in code has a lot of advantages over that including a builder and extensions:
```
                .AddTransforms(transformBuilderContext =>
                {
                    transformBuilderContext.AddXForwarded();
```

I'd re-use most of that directly except the TransformBuilderContext includes a Route and Cluster which don't make sense here. Maybe they could just be ignored?

```
            var transformBuilder = app.ApplicationServices.GetRequiredService<ITransformBuilder>();
            var transforms = transformBuilder.Create(transformBuilderContext =>
            {
                transformBuilderContext.AddXForwarded();
            });
```

**Comment by leastprivilege** [comment]
Yes, please make it so ;)

**Comment by karelz** [comment]
Triage: While interesting, it is rather advanced scenario. Only 1 ask so far, moving it to Backlog for now.

**Comment by leastprivilege** [comment]
Thanks for getting back! I do not think that `help wanted` is the right approach here, since this is a design question/issue.

Taking ProxyKit as an example, it was always "code first" and then one could layer configuration on top of the public APIs. So does the rest of ASP.NET Core.

YARP approaches this exactly the other way round and even actively prohibits some code first scenarios by making parts of the infrastructure pretty tied to config/hard to use or private.

I understand that this does not seem to be a priority for you (or at least right now) and I can't argue with that. But if your primary goal is to be config driven only, then there is a gap. This will make "direct proxying" scenarios always look like an afterthought.

In short term I guess it would help to make the `StructuredTransformer` public even if the API surface is not ideal. If that requires a PR, I can do that. 

Besides that, I would appreciate a discussion/statement about plans to make "code first" a priority (or not). Thanks!

**Comment by damianh** [comment]
> I would appreciate a discussion/statement about plans to make "code first" a priority (or not)

💯 to that please :)

I shut down ProxyKit, which was a code-first approach, fully expecting Yarp to support such. I'd consider rebooting it if you can definitively state that you're not interested in code-first ("advanced") scenarios, there definitely are users using that way. But please state one way or another - my time is precious too.

(Side note: as this is a design issue, I don't see how a `help-wanted` contribution could be successful).

**Comment by Tratcher** [comment]
There is an initial design proposal in https://github.com/microsoft/reverse-proxy/issues/801#issuecomment-788476716 that might be a sufficient starting point to iterate from in a PR.

As for the code-first discussion, can you take that to https://github.com/microsoft/reverse-proxy/discussions so we can keep this issue on topic?

**Comment by Tratcher** [comment]
For transparency on priorities, we're trying to wrap up 1.0 and are focusing on issues that would require breaking the current API surface. The APIs proposed here would be additive and could be introduce after 1.0.

**Comment by leastprivilege** [comment]
I had some time to experiment with this and copied `StructuredTransformer` to my project.

I can confirm that making this class public would unblock me (along with `RequestUtilities` but I think this was changed already).

Now I am running into the "cannot use init-only properties" problem which has been reported already. My library targets 3.1 and 5.0
